fontforge/python.c

```diff
@@ -1769,6 +1770,318 @@ Py_RETURN_NONE;
+		defcnt++;
+	    answer->name = getDictItemStringString(answer_spec, "name");
+	    tag = PyDict_GetItemString(answer_spec, "tag");
+	    if ( tag==NULL )
+		tag = PyDict_GetItemString(answer_spec, "name");
+	    if ( tag==NULL ) {
+		PyErr_Format(PyExc_TypeError, "askMulti: Answer for choice-type question '%s' must include either `name` or `tag`.");
+		Py_DECREF(answer_spec);
+		return false;
+	    }
+	    Py_INCREF(tag);
+	    answer->tag = tag;
+	    Py_DECREF(answer_spec);
+	}
+	if ( !qstn->multiple && defcnt>1 ) {
+	    PyErr_Format(PyExc_TypeError, "askMulti: Too many default answers for single-choice question '%s'.", qstn->label );
+	    return false;
+	}
+    } else {
+	qstn->dflt = getDictItemStringString(spec, "default");
+    }
+    return true;
+}
+
+static int multiDlgDecodeCategory(MultiDlgCategory *category, PyObject *spec, PyObject *tagdict) {
+    if ( (category->label = getDictItemStringString(spec, "category")) == NULL) {
+	PyErr_Format(PyExc_TypeError, "askMulti: Missing 'category' key in Category specification.");
+	return false;
+    }
+    PyObject *qstns_spec = PyDict_GetItemString(spec, "questions");
+    if ( qstns_spec==NULL || !PySequence_Check(qstns_spec) ) {
+	PyErr_Format(PyExc_TypeError, "askMulti: Category 'questions' key for '%s' is missing or is not a list.", category->label);
+	return false;
+    }
+    category->len = PySequence_Size(qstns_spec);
+    category->questions = calloc(category->len, sizeof(MultiDlgQuestion));
+    for (int i = 0; i<category->len; ++i) {
+        PyObject *qstn_spec = PySequence_GetItem(qstns_spec, i);
+        if ( !multiDlgDecodeQuestion(category->questions + i, qstn_spec, tagdict) ) {
+            return false;
+        }
+        Py_DECREF(qstn_spec);
+    }
+    return true;
+}
+
+PyObject *multiDlgExtractAnswers(MultiDlgSpec *dspec) {
+    PyObject *r = PyDict_New(), *k, *v, *vtuple;
+    int c, q, a, acnt, ai;
+
+    for ( c=0; c<dspec->len; ++c ) {
+	MultiDlgCategory *cspec = dspec->categories + c;
+	for ( q=0; q<cspec->len; ++q ) {
+	    MultiDlgQuestion *qspec = cspec->questions + q;
+	    if ( qspec->type==mdq_choice ) {
+		int acnt = 0;
+		for ( a=0; a<qspec->answer_len; ++a )
+		    if ( qspec->answers[a].is_checked )
+			acnt++;
+		if ( qspec->multiple )
+		    vtuple = PyTuple_New(acnt);
+		else
+		    assert( acnt<=1 );
+		ai = 0;
+		for ( a=0; a<qspec->answer_len; ++a )
+		    if ( qspec->answers[a].is_checked ) {
+			v = qspec->answers[a].tag;
+			Py_INCREF(v);
+			if ( qspec->multiple ) {
+			    PyTuple_SetItem(vtuple, ai, v);
+			    ai++;
+			} else
+			    break;
+		    }
+		if ( qspec->multiple )
+		    v = vtuple;
+	    } else {
+		if ( qspec->str_answer==NULL )
+		    v = Py_None;
+		else
+		    v = PyUnicode_FromString(qspec->str_answer);
+	    }
+	    k = qspec->tag;
+	    assert( !PyDict_Contains(r, k) );
+	    // XXX seems like v should have an extra reference at this point,
+	    // (because _SetItem() doesn't steal) but DECREFing v causes crashes
```

Comment:
I don't entirely understand what this means, but are you sure you don't need to increment the reference count on `Py_None`?

Reply:
That had been my impression but I guess that impression was wrong. I'll add the Py_None reference counting. 

`PyTuple_SetItem()` is documented as "stealing" a reference, hence the reference increment before the call. PyDict_SetItem() is documented as *not* stealing a reference: https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItem , which in my mind means it should take a reference, which means that I should be decrementing my own reference to it after the call, but that causes crashes. So there's something I'm missing.

Reply:
Ref count should be fixed